api/src/main/java/com/google/sps/SendNotifications.java

```diff
@@ -1,117 +1,159 @@
 import javax.mail.Session;
 import javax.mail.internet.*;
 
 import org.apache.commons.codec.binary.Base64;
-import org.springframework.boot.autoconfigure.SpringBootApplication;
-import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.PostMapping;
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
 public class SendNotifications {
+  static boolean firebaseInitialized = false;
 
-  public List<Email> gatherRecipients() throws IOException {
-    GoogleCredentials credentials = GoogleCredentials.getApplicationDefault();
-
-    System.out.println("Google Credentials created...");
-
-    String projectId = "phoebeliang-step";
-    FirebaseOptions options = new FirebaseOptions.Builder().setCredentials(credentials).setProjectId(projectId).build();
-    FirebaseApp.initializeApp(options);
-
-    System.out.println("App Initialized...");
+  public List<EmailCreation> gatherRecipients(HttpServletRequest request, HttpServletResponse response)
+    throws IOException {
 
     Firestore db = FirestoreClient.getFirestore();
-
-    String gameID = "dT6U1wQKR1rGJcEncyMS";
-    List<Email> emails = new ArrayList<>();
+    String gameID = request.getParameter("gameID");
+    EmailType emailType = EmailType.valueOf(request.getParameter("emailType").toUpperCase());
+    List<EmailCreation> emails = new ArrayList<>();
 
     DocumentReference game = db.collection("games").document(gameID);
-
+    CollectionReference players = db.collection("users");
     try {
       DocumentSnapshot gameDocSnap = game.get().get();
-      List<String> playerNames = (List<String>) gameDocSnap.get("players");
 
-      for(String player: playerNames) {
-        ApiFuture<QuerySnapshot> futurePlayers = db.collection("users").whereEqualTo("username", player).get();
-        List<QueryDocumentSnapshot> usersDocSnap = futurePlayers.get().getDocuments();
+      List<String> playerNames = ((List<String>) gameDocSnap.get("players"));
 
-        // Retrieves player information
-        Map<String,Object> playerInfo = usersDocSnap.get(0).getData();
-        String playerEmail = (String) playerInfo.get("email");
-        String playerName = (String) playerInfo.get("username");
+      for (String player : playerNames) {
+        DocumentSnapshot usersDocSnap = players.document(player).get().get();
 
-        System.out.println(playerEmail);
+        // Retrieves player information
+        String playerEmail = usersDocSnap.getString("email");
+        String playerName = usersDocSnap.getString("username");
 
         // Adding player to Email object
-        emails.add(new Email(new User(playerEmail, playerName)));
+        if (emailType == EmailType.START) {
+          emails.add(EmailCreation.startGameEmail(gameID, new User(playerEmail, playerName)));
+        } else if (emailType == EmailType.TURN) {
+          emails.add(EmailCreation.endGameEmail(gameID, new User(playerEmail, playerName)));
+        }
```

Comment:
In this specific case, have you considered using polymorphism and abstracting to enum methods? (An example was linked in the previous comment.)

Reply:
(btw this doesn't mean your code isn't valid; I just want to see evidence of design alternatives considered)

Reply:
I am assuming you mean the Payroll example? I read through each of the examples and after some thinking I figured that this was what you were getting at, since there's really only two different email types I thought this was the most straightforward